From 08d330741bbbc310d7df125ac6910661fec5cd5a Mon Sep 17 00:00:00 2001
From: Hiromitsu Yamasaki <hiromitsu.yamasaki.ym@renesas.com>
Date: Wed, 12 Sep 2018 17:52:14 +0900
Subject: [PATCH 240/909] i2c: rcar: Fix disable ICDMAER register

commit ece7b231072daaee86431233c80bfbde4740a2ca from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

Fixed a problem that the direction flag may be rewritten before
executing dma_unmap_single().

When I2C DMA is completed, rcar_i2c_dma_callback() is called as
CallBack from the DMAE driver(thread), and rcar_i2c_dma_unmap()
is called in it.

By disabling the DMA Enable Register(ICDMAER), rcar_i2c_dma_unmap()
enables interrupts for register settings(such as Master Control
Register(ICMCR)) in I2C and advances the I2C transfer sequence.

If a task switch occurs immediately after ICDMAER is disabled,
the I2C transfer sequence controlled by the I2C interrupt proceeds
and the DMA transfer of the next message is prepared(rcar_i2c_dma()).

DMA_NONE is set to the dma_direction flag when control returns to
rcar_i2c_dma_unmap() from task switching.

However, the dma_direction flag is set in preparing the DMA for
the next message.

By checking the dma_direction flag at the completion of DMA of
the next message due to the overwriting of the dma_direction flag
with DMA_NONE, BUG_ON() is displayed.

Reproduction test:
1. Add udelay(1) after disabling ICDMAER. (It is expected to
generate an interrupt of rcar_i2c_irq())

    void rcar_i2c_dma_unmap(struct rcar_i2c_priv *priv)
    {
        ...
        rcar_i2c_write(priv, ICDMAER, 0);
        udelay(1);                        <-- Add this line
        ...
        priv->dma_direction = DMA_NONE;
    }

2. Execute DMA transfer.
Performs DMA transfer of read or write. In the sample, write was used.

 $ i2cset -y -f 4 0x6a 0x01 0x01 0x02 0x03 0x04 0x05 0x06 0x07 0x08 i

3. A log message of BUG_ON () will be displayed.

Signed-off-by: Hiromitsu Yamasaki <hiromitsu.yamasaki.ym@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/i2c/busses/i2c-rcar.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/i2c/busses/i2c-rcar.c b/drivers/i2c/busses/i2c-rcar.c
index 9fbf8f7d74a4..53ed7eb6e5d2 100644
--- a/drivers/i2c/busses/i2c-rcar.c
+++ b/drivers/i2c/busses/i2c-rcar.c
@@ -370,9 +370,6 @@ static void rcar_i2c_dma_unmap(struct rcar_i2c_priv *priv)
 	struct dma_chan *chan = priv->dma_direction == DMA_FROM_DEVICE
 		? priv->dma_rx : priv->dma_tx;
 
-	/* Disable DMA Master Received/Transmitted */
-	rcar_i2c_write(priv, ICDMAER, 0);
-
 	dma_unmap_single(chan->device->dev, sg_dma_address(&priv->sg),
 			 sg_dma_len(&priv->sg), priv->dma_direction);
 
@@ -382,6 +379,9 @@ static void rcar_i2c_dma_unmap(struct rcar_i2c_priv *priv)
 		priv->flags |= ID_P_NO_RXDMA;
 
 	priv->dma_direction = DMA_NONE;
+
+	/* Disable DMA Master Received/Transmitted */
+	rcar_i2c_write(priv, ICDMAER, 0);
 }
 
 static void rcar_i2c_cleanup_dma(struct rcar_i2c_priv *priv)
-- 
2.17.1

